kolena/workflow/test_case.py

```diff
@@ -48,6 +52,90 @@
 from kolena.workflow.workflow import Workflow
 
 
+_BULK_PROCESSOR = contextvars.ContextVar("batch_loader")
+
+
+def check_edit(method):
+    @functools.wraps(method)
+    def is_editing(ref, *args, **kwargs):
+        if ref._bulk_editing:
+            raise IncorrectUsageError(f"Cannot invoke {method.__name__} when test-case is being bulk-processed")
+
+        return method(ref, *args, **kwargs)
+
+    return is_editing
+
+
+def _to_editor_data_frame(
+    edits: List[Tuple[TestSample, Optional[GroundTruth], bool]],
+    test_case_name: Optional[str] = None,
+) -> TestCaseEditorDataFrame:
+    records = [
+        (
+            test_case_name,
+            test_sample._data_type().value,
+            test_sample._to_dict(),
+            test_sample._to_metadata_dict(),
+            ground_truth._to_dict() if ground_truth is not None else None,
+            remove,
+        )
+        for test_sample, ground_truth, remove, in edits
+    ]
+    columns = [
+        "test_case_name",
+        "test_sample_type",
+        "test_sample",
+        "test_sample_metadata",
+        "ground_truth",
+        "remove",
+    ]
+    df = pd.DataFrame(records, columns=columns)
+    return TestCaseEditorDataFrame(validate_df_schema(df, TestCaseEditorDataFrame.get_schema(), trusted=True))
+
+
+@dataclass(frozen=True)
+class _BulkProcessor:
+    data: Dict[str, Tuple["TestCase", CoreAPI.SingleProcessRequest, TestCaseEditorDataFrame]] = dataclasses.field(
+        default_factory=dict,
+    )
+
+    def add(self, test_case: "TestCase", edit_data: CoreAPI.SingleProcessRequest, df: TestCaseEditorDataFrame) -> None:
+        if test_case.name in self.data:
+            raise IncorrectUsageError(f"Conflicting edits to test-case '{test_case.name}'")
+
+        with test_case._unfrozen():
+            test_case._enter_bulk_edit()
+        self.data[test_case.name] = (test_case, edit_data, df)
+
+    @staticmethod
+    def check_ignored(test_case: "TestCase", edit_data: CoreAPI.SingleProcessRequest, has_test_samples: bool) -> None:
+        if edit_data.description is not None and test_case.description != edit_data.description and not edit_data.reset:
+            log.warn(f"test case '{test_case.name}' already exists, not updating description when reset=False")
+        # there will be a false negative when target test-case is version 1
```

Comment:
Given the information client has, under bulk operation, it cannot tell the difference between
1. version-0 test case being updated, or
2. version-1 test case intended to updated without `reset`, thus ignored by server.

using `test_case.version > 1` will have false negative (no warning shown) for 2), and using `test_case.version > 0` will have false positive for 1).

Any preference?